docs/vm/proposal.md

```diff
@@ -0,0 +1,562 @@
+The compiler will translate semantic graph to IR, run optimization passes on the IR, then emit an IR suitable to be run by the VM.
+The VM will be defined in Go because our runtime is defined in Go.
+
+This plan allows us to gradually implement different instructions in the VM and continue to expand the scope of what can be run by the VM.
+The eventual goal is the removal of the interpreter and compiler from Go.
+
+At the end of development, the following will have been implemented:
+
+* Defined an SSA IR with a textual output.
+* Constructed function definitions.
+* Implemented function objects and call invocations.
+* Implemented closures as a part of function objects.
+* Constructed VM from SSA IR.
+* Ran VM program with function calls, closures, and plan execution operations.
+* Implement yield and table find.
+
+### Compiler
+
+The compiler, written in Rust, would be responsible for implementing the following:
+
+* Define an SSA IR with a textual output.
+* Translate semantic graph to IR.
+* Define function definitions that can be reified at runtime.
+
+The output of this would be a system roughly equivalent to what we have today.
+From there, we can use the debug output of the IR and use existing compiler optimization algorithms to implement optimization passes.
+The optimization passes would take IR as input and produce IR as output.
+
+### VM
+
+The VM, written in Go, would be responsible for implementing the instructions from the compiler.
+For the compiler to emit an instruction, it needs to be implemented in the VM.
+
+The VM is also responsible for constructing plan nodes when they are executed and passing them to the executor to be executed.
+The VM finishes execution when its main thread and all dispatcher threads have finished.
+
+## Design
+
+### Compiler
+
+The compiler would evaluate the semantic graph from the type system and produce an SSA IR.
+The IR would define various primitive operations that correspond to existing concepts.
+Looking at the IR would give a good indication of what the compiled flux script will do and the SSA form will allow easy construction of a control-flow graph.
+This will make it generally easier to perform introspection of the program along with making and debugging necessary transformations.
+
+The IR format will be representable textually to aid with debugging.
+There is no requirement that the underlying system interact with a textual IR.
+
+A sample of the textual output of an IR for a Flux program follows:
+
+    // y = 2
+    // addy = (x) => x + y
+    // addy(x: 1)
+    define (T: Addable) T @addy(%x: T, %y: T) {
+    add2:
+        %1 = add T %x, %y
+        ret T %1
+    }
+
+    define void @main() {
+        %1 = const int 2
+        %2 = func @add2, y: %1
+        %3 = const int 1
+        %4 = call int %2, x: %3
+        print %4
+        ret void
+    }
+
+Important concepts that are included in this output:
+
+* Constants
+* Function calls
+* Closures
+* Generics
```

Comment:
I'm personally of the opinion that we should front load the generics. I think it will be needed for the full version and, because it is both necessary and has a decent number of unknowns, I'd like it to be towards the beginning.

That's just my personal opinion on this. We could just have the runtime use dynamic types for everything to start.